### PR TITLE
Integrating Base/Module Classes with Codebase

### PR DESCRIPTION
**Pull request by itsfeas**
This pull request aims to merge the various classes, as well as the state machine code to master.

Currently to-do with this iteration of the code:

-Supply actual parameter values to StateMachine. If there is a need for functions to model bamocar input, please supply them. 

-Perform Unit testing of FlexCAN message FIFO setup to ensure that:

a) The code stops reading messages when there are no messages left to be read (StateMachine/main.cpp)

b) The FIFO buffer has been set up correctly, such that it can store messages until they can be read

-Perform Unit testing of BamocarInterface to ensure that unit conversions have been correctly supplied

-Unit Test StateMachine class position tracking. StateMachine should stop the run after 100m. This relies on a successful nav module. 

-Move the MessageMaster and prebuilt classes/headers to a better location

-Take a look at the test projects under 'Testing'. These were used for testing whether sensors/CANbus nodes had been successfully set up.

### CODE REVIEW

**Comment by itsfeas** [comment]
Do you guys mind taking a look at the motor_controller folder? 

@OkTerrific you might want to look at the MessageMaster class

File: StateMachine/src/main.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
+#include <Arduino.h>
+#include <Wire.h>
+#include <SPI.h>
+#include <FlexCAN_T4.h>
+
+#include "ChRt.h"
```
Comment (mbardwell) [inline StateMachine/src/main.cpp:6]: If threading proves to be error-prone we might remove it and try again in the future if we see the need. Feel free to leave it in for now

File: StateMachine/src/MessageMaster.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
+#ifndef _MSG_H_
+#define _MSG_H_
+
+#include <Arduino.h>
+#include <FlexCAN_T4.h>
+
+enum BamocarCanID
```
Comment (mbardwell) [inline StateMachine/src/MessageMaster.h:7]: Suggestion: have all possible CanID's in one `enum`

File: StateMachine/src/MessageMaster.h [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    uint8_t snd_msg_list[9][4];
+    uint8_t rcv_msg_list[3][5];
+    uint8_t nav_msg_list[3][2];
+
+  public:
+    const uint8_t bamocarTransmitLen = 3;
+    const uint8_t bamocarReceiveLen = 4;
+
+    MessageMaster(FlexCAN_T4<CAN0, RX_SIZE_256, TX_SIZE_16> &PrimaryBus, FlexCAN_T4<CAN1, RX_SIZE_256, TX_SIZE_16> &BamocarBus);
```
Comment (mbardwell) [inline StateMachine/src/MessageMaster.h:45]: again, for a more general `FlexCAN_T4`-based constructor, see [this](https://github.com/albertaloop/T_SWE_2019_2020/pull/31/files#diff-111717fa100e44ba15a71abc3bd81bcf27edabc806e998096c9bb65e7b06ec5fR55)

**Review by mbardwell — CHANGES_REQUESTED** [review]
A few general comments:
1. awesome work. Thanks for contributing!
2. we want to keep everything in a single `platformio.ini` file. That way, we can write libraries that are easily includable in a number of executables (ie. we will just need to add `#include "bleh.h"` to .cpp files). For this PR, Here, state machine, gpstest, etc... are all compiled using separate .ini files. I recommend merging them
3. Ideally, GPSTest would be added in a separate PR
4. There is existing code in the repo that does what the Sending/Recipient tests do. I recommend removing them from this PR

**Pull request closed without merging**